// @streamparser/json/dist/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenType = exports.TokenParserMode = exports.ParsedElementInfo = exports.ParsedTokenInfo = exports.JsonTypes = exports.utf8 = exports.TokenParserError = exports.TokenParser = exports.TokenizerError = exports.Tokenizer = exports.JSONParser = void 0;
var jsonparser_js_1 = require("./jsonparser.js");
Object.defineProperty(exports, "JSONParser", { enumerable: true, get: function () { return __importDefault(jsonparser_js_1).default; } });
var tokenizer_js_1 = require("./tokenizer.js");
Object.defineProperty(exports, "Tokenizer", { enumerable: true, get: function () { return __importDefault(tokenizer_js_1).default; } });
Object.defineProperty(exports, "TokenizerError", { enumerable: true, get: function () { return tokenizer_js_1.TokenizerError; } });
var tokenparser_js_1 = require("./tokenparser.js");
Object.defineProperty(exports, "TokenParser", { enumerable: true, get: function () { return __importDefault(tokenparser_js_1).default; } });
Object.defineProperty(exports, "TokenParserError", { enumerable: true, get: function () { return tokenparser_js_1.TokenParserError; } });
exports.utf8 = __importStar(require("./utils/utf-8.js"));
exports.JsonTypes = __importStar(require("./utils/types/jsonTypes.js"));
exports.ParsedTokenInfo = __importStar(require("./utils/types/parsedTokenInfo.js"));
exports.ParsedElementInfo = __importStar(require("./utils/types/parsedElementInfo.js"));
var stackElement_js_1 = require("./utils/types/stackElement.js");
Object.defineProperty(exports, "TokenParserMode", { enumerable: true, get: function () { return stackElement_js_1.TokenParserMode; } });
var tokenType_js_1 = require("./utils/types/tokenType.js");
Object.defineProperty(exports, "TokenType", { enumerable: true, get: function () { return __importDefault(tokenType_js_1).default; } });
